### PR TITLE
Decals/clustered

### PR DESCRIPTION
**Pull request by Paulm-Unity**

### CODE REVIEW

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/Decal.hlsl [hunk 1]
```diff
@@ -91,11 +91,24 @@ CBUFFER_START(UnityDecalParameters)
...
+TEXTURE2D(_DecalHTileTexture);
+
+uint _DecalCount;
+StructuredBuffer<DecalData> _DecalDatas;
+
+TEXTURE2D_ARRAY(_DecalAtlas);          
+SAMPLER(sampler_DecalAtlas);
+
+// define these specifically for decal normal map sampling to not be dependent on surface gradient
+#if defined(UNITY_NO_DXT5nm)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/Decal.hlsl]: We don't have the case of no DXT5nm for the platform of HD, you can safely use UnpackNormalmapRGorAG

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalData.hlsl [hunk 2]
```diff
@@ -10,17 +10,14 @@ void GetSurfaceData(float2 texCoordDS, float4x4 decalToWorld, out DecalSurfaceDa
...
 	surfaceData.baseColor = SAMPLE_TEXTURE2D(_BaseColorMap, sampler_BaseColorMap, texCoordDS.xy);
 	surfaceData.baseColor.w *= totalBlend;
 	surfaceData.HTileMask |= DBUFFERHTILEBIT_DIFFUSE;
 #endif
-	UVMapping texCoord;
-	ZERO_INITIALIZE(UVMapping, texCoord);
-	texCoord.uv = texCoordDS.xy;
 #if _NORMALMAP
-	surfaceData.normalWS.xyz = mul((float3x3)decalToWorld, SAMPLE_UVMAPPING_NORMALMAP(_NormalMap, sampler_NormalMap, texCoord, 1)) * 0.5f + 0.5f;
+	surfaceData.normalWS.xyz = mul((float3x3)decalToWorld, DECAL_UNPACK_NORMAL_FUNC(SAMPLE_TEXTURE2D(_NormalMap, sampler_NormalMap, texCoordDS))) * 0.5f + 0.5f;
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalData.hlsl]: as stated above, you can use directly UnpackNormalmapRGorAG here, make sense

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl [hunk 3]
```diff
@@ -2,21 +2,105 @@
...
+#else
+    int j = start + i;
+#endif
+    return _DecalDatas[j];
+}
+
+void ApplyBlendNormal(inout float4 dst, inout int matMask, float2 texCoords, int sliceIndex, int mapMask, float3x3 decalToWorld, float blend)
+{
+	float4 src;
+	src.xyz =  mul(decalToWorld, DECAL_UNPACK_NORMAL_FUNC(SAMPLE_TEXTURE2D_ARRAY(_DecalAtlas, sampler_DecalAtlas, texCoords, sliceIndex))) * 0.5f + 0.5f;
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: UnpackNormalmapRGorAG

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl [hunk 4]
```diff
@@ -2,21 +2,105 @@
...
+#ifdef _SURFACE_TYPE_TRANSPARENT	// forward transparent using clustered decals
+        uint decalCount, decalStart;
+		DBuffer0 = float4(0.0f, 0.0f, 0.0f, 1.0f);
+		DBuffer1 = float4(0.5f, 0.5f, 0.5f, 1.0f);
+		DBuffer2 = float4(0.0f, 0.0f, 0.0f, 1.0f);
+
+    #ifdef LIGHTLOOP_TILE_PASS
+        GetCountAndStart(posInput, LIGHTCATEGORY_DECAL, decalStart, decalCount);
+    #else
+        decalCount = _DecalCount;
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: does this mean you have implemented a brute force version with all the decal when we disable fptl (for debugging ) ? nice

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl [hunk 5]
```diff
@@ -2,21 +2,105 @@
...
+        decalStart = 0;
+    #endif
+		float3 positionWS = GetAbsolutePositionWS(posInput.positionWS);
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: hmm... I think here we could try to get a bit more precision if we pretranslate the decalData.worldToDecal with camera world pos (apply in GetAbsolutePositionWS), and thus no need to call GetAbsolutePositionWS.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs [hunk 6]
```diff
@@ -658,12 +661,16 @@ public override void Render(ScriptableRenderContext renderContext, Camera[] came
                     m_ReflectionProbeCullResults.Cull();
 
                     m_DbufferManager.vsibleDecalCount = 0;
-                    if (m_FrameSettings.enableDBuffer)
+                    using (new ProfilingSample(cmd, "DBufferPrepareDrawData", CustomSamplerId.DBufferPrepareDrawData.GetSampler()))
                     {
-                        m_DbufferManager.vsibleDecalCount = DecalSystem.instance.QueryCullResults();
-                        DecalSystem.instance.EndCull();
+                        if (m_FrameSettings.enableDBuffer)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: can you create a bool for DBuffer and one for cluster decal ?

**Comment by Paulm-Unity** [comment]
Its in DecalSystem.cs. Github hid the changes, because there were too many in that file, I had to manually expand them.

And actually now that I look at it, there is a bug there, it should not check for current index being -1, but rather set it to -1 if there is no texture.

 void UpdateTextureCache(CommandBuffer cmd)
            {
                if ((m_DiffuseTexIndex == -1) && (m_DiffuseTexture != null))
                {
                    m_DiffuseTexIndex = instance.TextureAtlas.FetchSlice(cmd, m_DiffuseTexture);
                }
                if ((m_NormalTexIndex == -1) && (m_NormalTexture != null))
                {
                    m_NormalTexIndex = instance.TextureAtlas.FetchSlice(cmd, m_NormalTexture);
                }
                if ((m_MaskTexIndex == -1) && (m_MaskTexture != null))
                {
                    m_MaskTexIndex = instance.TextureAtlas.FetchSlice(cmd, m_MaskTexture);
                }
            }

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs [hunk 7]
```diff
@@ -730,7 +737,7 @@ public void AllocResolutionDependentBuffers(int width, int height)
             }
         }
 
-        static Matrix4x4 WorldToCamera(Camera camera)
+        public static Matrix4x4 WorldToCamera(Camera camera)
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: Maybe I missed something in the PR. Where is it used as public?

**Comment by sebastienlagarde** [comment]
I merge the PR, but please if you can answer the various question :)

**Comment by Paulm-Unity** [comment]
I thought I answered everything except for why this needs to be public?
 public static Matrix4x4 WorldToCamera(Camera camera)

DecalSystem.cs needs it when computing decal bounding data for gpu light list generation.

If there are any other unanswered questions, please let me know, but I don't see any...

**Comment by sebastienlagarde** [comment]
Maybe you have commit your answer ? See on this page above. Various questions:)

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl (hunk 4 above)
Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: The structured buffer with all of the visible decals information always gets bound. I did test it with LIGHTLOOP_TILE_PASS disabled and it did work.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs (hunk 7 above)
Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: DecalSystem.cs needs it when computing decal bounding data for gpu light list generation.

If there are any other unanswered questions, please let me know, but I don't see any...

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl (hunk 5 above)
Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: ah right :)

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs (hunk 6 above)
Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: yes, I was forgetting that it only affect transparent. Fine

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs (hunk 7 above)
Comment (robbiesri) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: "Load Diff" strikes again!

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs [hunk 8]
```diff
@@ -714,9 +721,9 @@ public override void Render(ScriptableRenderContext renderContext, Camera[] came
 
                     RenderObjectsVelocity(m_CullResults, hdCamera, renderContext, cmd);
 
-                    RenderDBuffer(hdCamera, renderContext, cmd);
+                    RenderDBuffer(hdCamera, cmd);
 
-                    RenderGBuffer(m_CullResults, hdCamera, enableBakeShadowMask, renderContext, cmd);
+					RenderGBuffer(m_CullResults, hdCamera, enableBakeShadowMask, renderContext, cmd);
```
Comment (robbiesri) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: Nooo, tabs!

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs [hunk 9]
```diff
@@ -1690,7 +1697,9 @@ public bool PrepareLightsForGPU(CommandBuffer cmd, ShadowSettings shadowSettings
                     }
                 }
 
-                m_lightCount = m_lightList.lights.Count + m_lightList.envLights.Count;
+                m_lightList.bounds.AddRange(DecalSystem.m_Bounds);
+                m_lightList.lightVolumes.AddRange(DecalSystem.m_LightVolumes);
+                m_lightCount = m_lightList.lights.Count + m_lightList.envLights.Count + DecalSystem.m_DecalDatasCount;
```
Comment (robbiesri) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: Hrm, I'm going to multi-plex this soon with two lists (one per stereo eye).  Maybe there's a more elegant way of managing these combined lists.  The code here is getting a bit hard to follow. Just a quick read of this code is kind of unobvious (the light count is lights.count, envLights.count and DecalDatasCount.  Maybe we should rename m_lightCount....

Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: Yeah, decals are treated as lights here for the purposes of GPU culling, if we are going to rename though, then to be consistent we should probably rename everything that deals with light culling, because now its not just lights.
